File: engine/tests/fixtures/comprehensive.java
import java.util.List;
import java.util.Map;
import java.util.Set;

interface Solvable<T> {
  Map<T, Integer> solve(int limit);
}

enum Mode { FAST, SAFE }

final class Solver extends BaseSolver implements Solvable<Integer> {
  static final int VERSION = 1;
  private final List<Integer> values;

  Solver(List<Integer> values) {
    this.values = values;
  }

  @Override
  public Map<Integer, Integer> solve(int limit) {
    Map<Integer, Integer> seen = new java.util.HashMap<>();
    Set<Integer> unique = new java.util.HashSet<>(values);
    List<Integer> prefix = values.subList(0, limit);
    for (int value : prefix) {
      if (value > 0) seen.put(value, solve(limit - 1).size());
    }
    while (limit > 0) limit -= 1;
    return seen;
  }
}

File: engine/tests/fixtures/comprehensive.rs
use std::collections::{HashMap, HashSet};

enum Mode {
    Fast,
    Safe,
}

trait Solvable {
    fn solve(&self, limit: usize) -> HashMap<i64, usize>;
}

struct Solver {
    values: Vec<i64>,
}

impl Solver {
    const VERSION: i64 = 1;

    fn new(values: Vec<i64>) -> Self {
        Self { values }
    }
}

impl Solvable for Solver {
    fn solve(&self, mut limit: usize) -> HashMap<i64, usize> {
        let mut seen = HashMap::new();
        let unique = HashSet::from_iter(self.values.clone());
        let prefix = self.values[..limit].to_vec();
        for value in &prefix {
            if *value > 0 {
                seen.insert(*value, self.solve(limit - 1).len());
            }
        }
        while limit > 0 {
            limit -= 1;
        }
        seen
    }
}
